### PR TITLE
Add BearSSL extra

### PR DESCRIPTION
**Pull request by meriac**
# Changes
- Based on #284 by @stschake, resolved merge conflicts
- Update subrepository to latest BearSSL version
- Use full certificate chain verification
- Add Let's-Encrypt root certificate (IdenTrust cross-signed) to support howsmyssl's trust anchor
- Add workaround for missing date & time (use time of compilation for SSL for now as start date)
- Add updated certificates for tls_server_bearssl example from BearSSL server example,
  previous ones were expired
- Verified both examples to work

# Testing
## Client verification
Example output on serial console at 115200 baud:
```
ESP-Open-SDK ver: 0.0.1 compiled @ Feb 13 2017 02:55:12
phy ver: 273, pp ver: 8.3

SDK version:0.9.9
mode : sta(a0:20:a6:00:ff:ab)
add if0
scandone
add 0
aid 3
cnt 

connected with OpenBeacon.org, channel 3
dhcp client start...
ip:10.254.1.247,mask:255.255.0.0,gw:10.254.0.1
Initializing BearSSL... done.
Connected
HTTP/1.1 200 OK
Content-Length: 1727
Access-Control-Allow-Origin: *
Connection: close
Content-Type: application/json
Date: Mon, 13 Feb 2017 03:05:39 GMT
Strict-Transport-Security: max-age=631138519; includeSubdomains; preload

{"given_cipher_suites":["TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256","TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256","TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256","TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256","TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384","TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384","TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256","TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256","TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384","TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384","TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA","TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA","TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA","TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA","TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256","TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256","TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384","TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384","TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256","TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256","TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384","TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384","TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA","TLS_ECDH_RSA_WITH_AES_128_CBC_SHA","TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA","TLS_ECDH_RSA_WITH_AES_256_CBC_SHA","TLS_RSA_WITH_AES_128_GCM_SHA256","TLS_RSA_WITH_AES_256_GCM_SHA384","TLS_RSA_WITH_AES_128_CBC_SHA256","TLS_RSA_WITH_AES_256_CBC_SHA256","TLS_RSA_WITH_AES_128_CBC_SHA","TLS_RSA_WITH_AES_256_CBC_SHA","TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA","TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA","TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA","TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA","TLS_RSA_WITH_3DES_EDE_CBC_SHA"],"ephemeral_keys_supported":true,"session_ticket_supported":false,"tls_compression_supported":false,"unknown_cipher_suite_supported":false,"beast_vuln":false,"able_to_detect_n_minus_one_splitting":false,"insecure_cipher_suites":{},"tls_version":"TLS 1.2","rating":"Improvable"}

free heap pre  = 12836
free heap post = 13128
successes = 1 failures = 0
10...
9...
8...
7...
6...
5...
4...
3...
2...
1...
0...
Starting again!
```
## Server verification
Use OpenSSL s_client to verify connection on command line:
```bash
> openssl s_client -host 10.254.1.247 -port 800 -verify_hostname localhost
```
Example output on the terminal:
```
CONNECTED(00000003)
depth=0 C = CA, CN = localhost
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = CA, CN = localhost
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/C=CA/CN=localhost
   i:/C=CA/CN=Intermediate CA
 1 s:/C=CA/CN=Intermediate CA
   i:/C=CA/CN=Root
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDPDCCAiSgAwIBAgIUWNq6Ns3toNpcEDNzjgxkknmSrwMwDQYJKoZIhvcNAQEL
BQAwJzELMAkGA1UEBhMCQ0ExGDAWBgNVBAMTD0ludGVybWVkaWF0ZSBDQTAeFw0x
MDAxMDEwMDAwMDBaFw0zNzEyMzEyMzU5NTlaMCExCzAJBgNVBAYTAkNBMRIwEAYD
VQQDEwlsb2NhbGhvc3QwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDU
eh0nuis6Z7KRavvng0TK7Rx1rd1Ng2LWqmiVsiQhexWuKploFe1m8LhY59P1LsbZ
Kl7nDi7n/GdZwMhhfUukb92f2ciFh2THuhoPKdSWqHiaa2IgqTLQ7qmMKGFHolAq
Y/Yh3trY1fB/xQCCcOajv1yJJ09RkncDw7DMLjvsI/IvU0GviZP/0oCxQ5fe1hmg
khJ6PWZ54cG84Xdwoos9RoRTP+ROQkE3kh4f/Tiz9++HOYDTVs/04BPeZLBypAOE
xEHtb/o+4soEINLX3CyCK3ribaEcSNvPiU80lz0oqFPa58HhcxWjMHZ/jyNCFD1R
NNJarTyby8j+f26OQPO9AgMBAAGjZjBkMB8GA1UdIwQYMBaAFMUBrXzmY8mcF1/F
oqfhUF/o9ajGMB0GA1UdDgQWBBTFAa185mPJnBdfxaKn4VBf6PWoxjAMBgNVHRMB
Af8EAjAAMBQGA1UdEQQNMAuCCWxvY2FsaG9zdDANBgkqhkiG9w0BAQsFAAOCAQEA
cbNdIcIO19DG+Epzh00iAifQx/j9Gm1iWIIIdiAHwEiS8+mYWusNTlaVY2hNq9QA
duA3zwsRYVlc3valFFnZJZ9Z2dNehqwdpiwyQhkyE0ALVM1nJra9tJakyh9/N9ao
des6gVEwuflKAW/R1u1P3z8wYAZnko5hhV8atYyzD2Gp+t9dxGQA6oexM199y6OF
JG4sZTvqcz+G0/3o5ALGYWomF1IBJVx/qM5pH6xhLLcEr/2kepnLJhVM/3TUcwxX
DCbr1yrcXMNBu8Lzzha9jnv76d+rIQ2Rs43Yz8j0SbnQ4xZwP7Pe1Acl+kZEUolN
icjiyrUzf8chvSjv/mZ0Aw==
-----END CERTIFICATE-----
subject=/C=CA/CN=localhost
issuer=/C=CA/CN=Intermediate CA
---
No client certificate CA names sent
Peer signing digest: SHA256
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 2143 bytes and written 327 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: EF82CE24AE7A7C8ED88A98A38F395136F997C9A0C41D7A43D19330B54CA0840F
    Session-ID-ctx: 
    Master-Key: E269116E9FE6F12862FE2C6A45C9D48BFC2B426367A9604B0329660543E5439DCE5E850FCEB0EE53C7ECBF646CDF1F00
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Start Time: 1486951219
    Timeout   : 300 (sec)
    Verify return code: 21 (unable to verify the first certificate)
---
O hai, client 10.254.0.8:52404
Free heap size is 16676 bytes
closed
```

### CODE REVIEW

**Comment by UncleRus** [comment]
Thank you, nice contribution!
As @sheinz already said in #284 it's always good to have options to choose from :)

**Comment by meriac** [comment]
@UncleRus I extended the timer code to [continuously progress time](https://github.com/SuperHouse/esp-open-rtos/pull/340/commits/c060a9c82a9558a51221b43ae9e9843f6d9b3bcd#diff-6314773a2e2994cb71e821eebc6bad48R233). From a functional perspective I believe the code is ready to be merged now.

File: examples/http_get_bearssl/http_get_bearssl.c [hunk 1]
```diff
@@ -201,6 +229,15 @@ void http_get_task(void *pvParameters)
         br_sslio_init(&ioc, &sc.eng, sock_read, &fd, sock_write, &fd);
         printf("done.\r\n");
 
+        /* FIXME: set date & time using epoch time precompiler flag for now */
+        some_time = CONFIG_EPOCH_TIME + (xTaskGetTickCount()/configTICK_RATE_HZ);
+        xc.days = (some_time / CLOCK_SECONDS_PER_DAY) + 719528;
+        xc.seconds = some_time % CLOCK_SECONDS_PER_DAY;
```
Comment (meriac) [inline examples/http_get_bearssl/http_get_bearssl.c]: @UncleRus I extended the timer code to [continuously progress time](https://github.com/SuperHouse/esp-open-rtos/pull/340/commits/c060a9c82a9558a51221b43ae9e9843f6d9b3bcd#diff-6314773a2e2994cb71e821eebc6bad48R233). From a functional perspective I believe the code is ready to be merged now.

File: examples/http_get_bearssl/http_get_bearssl.c [hunk 2]
```diff
@@ -201,6 +229,15 @@ void http_get_task(void *pvParameters)
         br_sslio_init(&ioc, &sc.eng, sock_read, &fd, sock_write, &fd);
         printf("done.\r\n");
 
+        /* FIXME: set date & time using epoch time precompiler flag for now */
+        provisional_time = CONFIG_EPOCH_TIME + (xTaskGetTickCount()/configTICK_RATE_HZ);
+        xc.days = (provisional_time / CLOCK_SECONDS_PER_DAY) + 719528;
+        xc.seconds = provisional_time % CLOCK_SECONDS_PER_DAY;
```
Comment (meriac) [inline examples/http_get_bearssl/http_get_bearssl.c]: @UncleRus I extended the timer code to continuously progress time. From a functional perspective I believe the code is ready to be merged now.

**Pull request merged**